### PR TITLE
Add test for SPI

### PR DESCRIPTION
**Pull request by tweksteen**
Add a test for the SPI util info command. Emulate SPIBAR and verify that BC and FRAP are correctly filled.

### CODE REVIEW

**Comment by timevortex** [comment]
From looking at this and searching around I don't see this being called by anything at this point. How/where/when will this be used?

Can you also please add a generous amounts of comments throughout to make it easier to read the code :) I am really not sure what you're trying to achieve and why!

**Comment by af3556** [comment]
As a general guide, if a unit test needs comments to understand then it probably needs to be simplified. i.e. unit tests should be so small and "obvious" that comments are not required.

Mind you, describing the meaning of magic numbers is probably not a terrible idea.

File: source/tool/tests/software/test_util.py [hunk 1]
```diff
@@ -149,3 +149,47 @@ def read_phys_mem(self, pa_hi, pa_lo, length):
...
+                                                               function,
+                                                               address, size)
+                elif (bus, device, function, address) == (
+                      0, 0x1F, 0, 0xF0):
+                    return self.RCBA_ADDR
+                elif (bus, device, function, address) == (
+                      0, 0x1F, 0, 0xDC):
+                    return 0xDEADBEEF
+                else:
+                    raise Exception("Unexpected PCI reading")
```
Comment (af3556) [inline source/tool/tests/software/test_util.py]: "Unexpected PCI BDF"

File: source/tool/tests/software/test_util.py [hunk 2]
```diff
@@ -149,3 +149,47 @@ def read_phys_mem(self, pa_hi, pa_lo, length):
...
+                if pa == self.FREG0:
+                    return 0x11111111
+                elif pa == self.FREG0 + 4:
+                    return 0x22222222
+                elif pa == self.FRAP:
+                    return 0xEEEEEEEE
+                elif pa >= self.SPIBAR_ADDR and pa < self.SPIBAR_END:
+                    return 0x0
+                else:
+                    raise Exception("Incorrect range")
```
Comment (af3556) [inline source/tool/tests/software/test_util.py]: "Unexpected address"

**Comment by tweksteen** [comment]
Comments added. Currently there is no documentation on how to run the tests. I'll add that in a separate commit. For now, within the source/tool directory:

`python -m unittest discover tests/software/`

File: source/tool/tests/software/test_util.py [hunk 3]
```diff
@@ -149,3 +149,47 @@ def read_phys_mem(self, pa_hi, pa_lo, length):
...
+
+            def write_mmio_reg(self, pa, size, value):
+                if pa < self.SPIBAR_ADDR or pa > self.SPIBAR_END:
+                    raise Exception("Writing out of SPIBAR")
```
Comment (af3556) [inline source/tool/tests/software/test_util.py]: "Write to outside of SPI range"

**Comment by af3556** [comment]
lgtm

**Comment by timevortex** [comment]
LGTM

Thanks. Looks great.

**Pull request closed without merging**